### PR TITLE
Extend DoubleQDQPairsRemover to handle sequences that end in duplicate DQ nodes

### PR DESCRIPTION
**Pull request by adrianlizarraga**
### Description
Extend the DoubleQDQPairsRemover optimizer to also handle sequences that end in duplicate DQ nodes.

For example, the following sequence:
```
 Q1 --> DQ1 --> Q2 --+--> DQ2
                     |
                     +--> DQ2'
```
Is now simplified to:
```
 Q1 ---+--> DQ2
       |
       +--> DQ2'
```

### Motivation and Context
The EnsureUniqueDQNodeUnits pass may add duplicate DQ nodes to ensure valid QDQ node units. The DoubleQDQPairsRemover should still be able to remove unnecessary QDQ ops if the target sequence ends in duplicate DQ nodes.

### CODE REVIEW

File: onnxruntime/core/optimizer/double_qdq_pairs_remover.cc
Comment (adrianlizarraga) [inline onnxruntime/core/optimizer/double_qdq_pairs_remover.cc:200]: This check is new. Need to make sure that both (Q -> DQ) use the same quantization type. This reduction is invalid otherwise.

File: onnxruntime/core/optimizer/double_qdq_pairs_remover.cc
Comment (yuslepukhin) [inline onnxruntime/core/optimizer/double_qdq_pairs_remover.cc:206]: InlinedVector everywhere. For 64-bit ptrs it can fit up to 6 wo memory allocations

**Review by yuslepukhin — CHANGES_REQUESTED** [review]

File: onnxruntime/test/optimizer/graph_transform_test_builder.h
Comment (adrianlizarraga) [reply onnxruntime/test/optimizer/graph_transform_test_builder.h:206]: I tried doing this, but this change would require modifying a large number of tests across various files. The main issue is that passing an initializer list, such as `{1, 2, 3}`, as an argument for a `gsl::span` parameter fails to compile. Would need to modify all call sites to pass in `AsSpan(std::array{1, 2, 3})` instead.

Although I think this would be a good cleanup task, I would prefer to keep this PR more focused.

File: onnxruntime/core/optimizer/double_qdq_pairs_remover.cc
Comment (edgchen1) [inline onnxruntime/core/optimizer/double_qdq_pairs_remover.cc:186]: nit: naming suggestion - q1_quant_type and q2_quant_type

File: onnxruntime/test/optimizer/graph_transform_test_builder.cc
Comment (edgchen1) [inline onnxruntime/test/optimizer/graph_transform_test_builder.cc:29]: would using gsl::span::as_bytes be simpler?

File: onnxruntime/test/optimizer/graph_transform_test_builder.h
Comment (adrianlizarraga) [reply onnxruntime/test/optimizer/graph_transform_test_builder.h]: Updated

File: onnxruntime/test/optimizer/graph_transform_test_builder.cc
Comment (adrianlizarraga) [reply onnxruntime/test/optimizer/graph_transform_test_builder.cc]: Updated

File: onnxruntime/core/optimizer/double_qdq_pairs_remover.cc
Comment (adrianlizarraga) [reply onnxruntime/core/optimizer/double_qdq_pairs_remover.cc]: Updated

File: onnxruntime/core/optimizer/double_qdq_pairs_remover.cc
Comment (adrianlizarraga) [reply onnxruntime/core/optimizer/double_qdq_pairs_remover.cc]: Updated

File: onnxruntime/core/optimizer/double_qdq_pairs_remover.cc
Comment (adrianlizarraga) [reply onnxruntime/core/optimizer/double_qdq_pairs_remover.cc]: Thanks. I missed this. Updated.

File: onnxruntime/test/optimizer/graph_transform_test_builder.cc
Comment (adrianlizarraga) [reply onnxruntime/test/optimizer/graph_transform_test_builder.cc]: Yes, good call. I updated this.

**Review by HectorSVC — APPROVED** [review]
:shipit:

**Comment by adrianlizarraga** [comment]
@yuslepukhin I addressed your comments. Could you please take another look? Merging is blocked on Github until you approve that your requested changes have been addressed.

**Review by jywu-msft — APPROVED** [review]

**Pull request merged**
